File: bindings/java/src/main/java/com/mystilink/liuyao/LiuyaoCalculator.java
package com.mystilink.liuyao;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/**
 * Invokes the liuyao CLI and returns JSON stdout.
 */
public final class LiuyaoCalculator {
    private LiuyaoCalculator() {}

    public static String resolveCli() {
        String env = System.getenv("MYSTILINK_LIUYAO_CLI");
        if (env != null && !env.isBlank()) {
            return env;
        }
        return "liuyao";
    }

    public static String run(String... args) throws Exception {
        List<String> cmd = new ArrayList<>();
        cmd.add(resolveCli());
        for (String a : args) {
            cmd.add(a);
        }
        ProcessBuilder pb = new ProcessBuilder(cmd);
        pb.redirectErrorStream(false);
        Process proc = pb.start();
        StringBuilder stdout = new StringBuilder();
        StringBuilder stderr = new StringBuilder();
        try (BufferedReader out = new BufferedReader(
                new InputStreamReader(proc.getInputStream(), StandardCharsets.UTF_8));
             BufferedReader err = new BufferedReader(
                new InputStreamReader(proc.getErrorStream(), StandardCharsets.UTF_8))) {
            String line;
            while ((line = out.readLine()) != null) {
                stdout.append(line).append('\n');
            }
            while ((line = err.readLine()) != null) {
                stderr.append(line).append('\n');
            }
        }
        int code = proc.waitFor();
        if (code != 0) {
            String msg = stderr.length() > 0 ? stderr.toString() : stdout.toString();
            if (msg.isBlank()) {
                msg = "liuyao exited with code " + code;
            }
            throw new IllegalStateException(msg.trim());
        }
        return stdout.toString();
    }

    public static String cast(Integer seed, String throwsCsv) throws Exception {
        List<String> args = new ArrayList<>();
        args.add("cast");
        if (throwsCsv != null && !throwsCsv.isEmpty()) {
            args.add("--throws");
            args.add(throwsCsv);
        } else if (seed != null) {
            args.add("--seed");
            args.add(String.valueOf(seed));
        }
        return run(args.toArray(new String[0]));
    }

    public static String version() throws Exception {
        return run("version");
    }
}

File: bindings/csharp/LiuyaoCalculator.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;

namespace Mystilink.Liuyao
{
    /// <summary>
    /// Invokes the liuyao CLI and returns JSON stdout.
    /// </summary>
    public static class LiuyaoCalculator
    {
        public static string ResolveCli()
        {
            var env = Environment.GetEnvironmentVariable("MYSTILINK_LIUYAO_CLI");
            return string.IsNullOrWhiteSpace(env) ? "liuyao" : env;
        }

        public static string Run(params string[] args)
        {
            var psi = new ProcessStartInfo
            {
                FileName = ResolveCli(),
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true,
                StandardOutputEncoding = Encoding.UTF8,
                StandardErrorEncoding = Encoding.UTF8,
            };
            foreach (var a in args)
            {
                psi.ArgumentList.Add(a);
            }

            using var proc = Process.Start(psi)
                ?? throw new InvalidOperationException("Failed to start liuyao");
            string stdout = proc.StandardOutput.ReadToEnd();
            string stderr = proc.StandardError.ReadToEnd();
            proc.WaitForExit();
            if (proc.ExitCode != 0)
            {
                var msg = string.IsNullOrWhiteSpace(stderr) ? stdout : stderr;
                throw new InvalidOperationException(
                    string.IsNullOrWhiteSpace(msg)
                        ? $"liuyao exited with code {proc.ExitCode}"
                        : msg.Trim());
            }
            return stdout;
        }

        public static string Cast(int? seed = null, string? throws = null)
        {
            var args = new List<string> { "cast" };
            if (!string.IsNullOrEmpty(throws))
            {
                args.Add("--throws");
                args.Add(throws);
            }
            else if (seed.HasValue)
            {
                args.Add("--seed");
                args.Add(seed.Value.ToString());
            }
            return Run(args.ToArray());
        }

        public static string Version()
        {
            return Run("version");
        }
    }
}
